### DISCUSSION TITLE
Any tutorial on testing with Pytest?

### DISCUSSION THREAD

**Discussion by hyochan** [RobertCraigie/prisma-client-py#432] [Q&A]
Hi. Thanks for the great work 💯 

I am wondering if there is any great example of testing APIs with `prismy-client-py`.

It'd be also great if there is some documents.

**Comment by RobertCraigie** [comment]
Thank you! :)

> I am wondering if there is any great example of testing APIs with prisma-client-py.

There aren't any "official examples" on how to write tests for an application that uses Prisma Client Python but the internal pytest config would be a good reference. Notable functions include `client_fixture`, `patch_prisma_fixture`, `setup_client_fixture` and `cleanup_client`. https://github.com/RobertCraigie/prisma-client-py/blob/c37ea68219f64df8da01b642bf1a0781987217c4/tests/conftest.py

In regards to having a clean database for each test there are a couple options, copy the `cleanup_client` function mentioned above, or use something specific to your database provider, e.g. this function was shared by a user in the discord: 

![Screen_Shot_2022-02-10_at_2 37 23_pm](https://user-images.githubusercontent.com/23125036/174573737-9adb5b28-1ad4-46e0-8050-c1b22c41518f.png)

And here's their full contest: https://cdn.discordapp.com/attachments/933860923117043718/941178322484883466/conftest.py

Having official documentation on this is a very good idea and something that should be included in the future.

**Reply by hyochan** [reply]
@RobertCraigie Thanks for your kindness. Would you mind giving minimal help on setting up `conftest.py` in [repo pr#4](https://github.com/prisma-korea/prisma-fastapi/pull/4)?

I am not quite sure how to set it up yet 🤔
<img width="360" alt="Screen Shot 2022-06-22 at 10 06 09 PM" src="https://user-images.githubusercontent.com/27461460/175036297-11cce551-bb09-4a94-9abe-64463b6fecb4.png">

**Reply by RobertCraigie** [reply]
Of course, I'm more than happy to help :)

By default, `pytest-asyncio` uses a function scoped event loop which means the loop gets torn down after each test invocation. This is inherently incompatible with using a session scoped `Prisma` instance (which you should be doing as creating the connection pool is expensive). In order to disable the default behaviour you have to define your own `event_loop` fixture that is session scoped. See this file for an example: https://github.com/RobertCraigie/prisma-client-py/blob/c37ea68219f64df8da01b642bf1a0781987217c4/tests/conftest.py#L42-L44
